File: Binary Trees and Binary Search Trees/1448. Count Good Nodes in Binary Tree/1448. Count Good Nodes in Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function (root) {
  const getGoodNodes = (root, curr_max) => {
    if (!root) return 0;

    if (root.val >= curr_max) {
      return (
        1 +
        getGoodNodes(root.left, root.val) +
        getGoodNodes(root.right, root.val)
      );
    }
    return (
      getGoodNodes(root.left, curr_max) + getGoodNodes(root.right, curr_max)
    );
  };

  return getGoodNodes(root, -Infinity);
};

// Time: O(n)
// Space: O(n)

File: Binary Trees and Binary Search Trees/1448. Count Good Nodes in Binary Tree/1448. Count Good Nodes in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def getGoodNodes(root, curr_max):
            if not root:
                return 0

            if root.val >= curr_max:
                return 1 + getGoodNodes(root.left, root.val) + getGoodNodes(root.right, root.val)

            return getGoodNodes(root.left, curr_max) + getGoodNodes(root.right, curr_max)
        
        return getGoodNodes(root, -float("inf"))

# Time: O(n)
# Space: O(n)